### DISCUSSION TITLE
Class and Struct can have the same names

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/csharplang#6341] [Language Ideas] [answered]
The compiler can choose whether it's a struct or a class when you call their constructor and etc

**Comment by HaloFour** [comment, marked as answer]
How do you propose the compiler do this?  Types need to be defined as metadata in the assembly, and they must either extend from `System.ValueType` or not.  It's not for the compiler to decide what a type is, nor can both a struct and a class have the same name within an assembly.

**Reply by Xyncgas** [reply]
Static analysis, we figure everything out as if we have a human trying to figure out it's a struct or a class, then codes will be compiled naturally

**Reply by Xyncgas** [reply]
And the best part is, depending on different places in your code, the same thing can be either a class or a struct out of reasons like optimizations

**Reply by Xyncgas** [reply]
So in the future instead of having class, struct, we just have one single thing that we put our members inside, and the moment you use value type features in the definition it's automatically a struct and you can write another thing with the same name that doesn't have any of the chrematistic of a value type and now it can be 'any type', it's all specialized and unified, this allows us write code that's cleaner because imagine instead of having ValueTask and Task you just have one single type Task and it will automatically be a value type or a reference type depending on things

**Reply by Xyncgas** [reply]
Because sometimes it's very easy to see that things are just supposed to be a class, and sometimes it's easy to see that things should better be a struct, having the compiler do it for you is absolutely the best thing in this world

**Reply by Xyncgas** [reply]
One example for you to imagine is you have a hidden struct that is functionally the same as a class even though that class has inheritance and might need to cast back and stuffs, don't worry, just stuff all the members of the class structure in that value type struct for now, and by default this 'thing' is a class, but during compilation we realized that new constructor can be replaced with that of the value type's, for performance reasons

**Reply by CyrusNajmabadi** [reply]
> Static analysis, we figure everything out as if we have a human trying to figure out it's a struct or a class, then codes will be compiled naturally

There's no way to know when compiling.  It depends on consumers you don't have access to.

**Reply by CyrusNajmabadi** [reply]
> and it will automatically be a value type or a reference type depending on things

This is not a virtue.  This means you now cannot tell easily what is going on, which is the last thing you want with systems designed to be explicit for performance reasons.

**Reply by CyrusNajmabadi** [reply]
>  it's easy to see that things should better be a struct, having the compiler do it for you is absolutely the best thing in this world

If it's easy to see, then it would be better for you to write an analyzer that just tells you this.  If you manually picked the wrong type, the analyzer can tell you to change it.  But now it never changes between one or the other randomly.
